st2common/st2common/models/api/base.py

```diff
@@ -71,7 +70,22 @@ def __json__(self):
         return vars(self)
 
     def validate(self):
-        VALIDATOR(getattr(self, 'schema', {})).validate(vars(self))
+        """
+        Perform validation and return cleaned object on success.
+        """
+        schema = getattr(self, 'schema', {})
+        attributes = vars(self)
+
+        cleaned = util_schema.validate(instance=attributes, schema=schema,
+                                       cls=util_schema.CustomValidator, use_default=True)
+
+        # Note: For backward compatibility reasons, we re-assign attributes one the current
```

Comment:
Yes. destructive interfaces like this are problematic and lead to surprising behaviors. It would be great if we move away to something safer for sure.